### PR TITLE
[WIP] Extended aamp Algorithms to p-norm and added unit test cases

### PR DESCRIPTION
**Pull request by vaish-muk**
code for  https://github.com/TDAmeritrade/stumpy/issues/251

- Modified aamp.py to generalise for p-norm ,calculated absolute of difference and raised it to the power of p ,added condition to use np.sqrt when p = 2 and np.power with 1/p for p!=2

- Modified naive.py to compare minkowski distance to generalise for p!=2 as well, minkowski distance = euclidian distance for p=2 hence works for base case as  well

- Modified stump.py with argument p_norm and default value as 2 to handle parameter comparison while testing

-  Added unit test cases for p=1,2,3 and verified results

### CODE REVIEW

**Comment by seanlaw** [comment]
@vaish-muk Thank you. Please allow me some time to review

**Review by seanlaw — COMMENTED** [review]
@vaish-muk Overall this is a good start. In order for this PR to be complete, p-norm also needs to be added to:

`ammped.py`
`gpu_aamp.py`
`maamp.py`

If it is not possible to add it to all of these places then we should avoid adding this feature altogether.

File: tests/test_aamp.py [hunk 1]
```diff
@@ -39,6 +39,46 @@ def test_aamp_self_join(T_A, T_B):
...
+    ref_mp = naive.aamp(T_A, m, T_B=T_B, p_norm=1.0)
+    comp_mp = aamp(T_A, m, T_B, ignore_trivial=False, p_norm=1.0)
+    naive.replace_inf(ref_mp)
+    naive.replace_inf(comp_mp)
+    npt.assert_almost_equal(ref_mp, comp_mp)
+    comp_mp = aamp(pd.Series(T_A), m, pd.Series(T_B), ignore_trivial=False, p_norm=1.0)
+    naive.replace_inf(comp_mp)
+    npt.assert_almost_equal(ref_mp, comp_mp)
+
+
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply tests/test_aamp.py]: added to latest commit

File: tests/naive.py [hunk 2]
```diff
@@ -204,7 +204,7 @@ def stump(T_A, m, T_B=None, exclusion_zone=None):
     return result
 
 
-def aamp(T_A, m, T_B=None, exclusion_zone=None):
+def aamp(T_A, m, T_B=None, exclusion_zone=None, p_norm=2):
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply tests/naive.py]: added to latest commit

File: stumpy/stump.py [hunk 3]
```diff
@@ -606,6 +608,7 @@ def stump(T_A, m, T_B=None, ignore_trivial=True, normalize=True):
         T_B_subseq_isconstant,
         diags,
         ignore_trivial,
+        p_norm,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/stump.py]: added to latest commit

File: stumpy/stump.py [hunk 4]
```diff
@@ -433,7 +435,7 @@ def _stump(
 
 
 @core.non_normalized(aamp)
-def stump(T_A, m, T_B=None, ignore_trivial=True, normalize=True):
+def stump(T_A, m, T_B=None, ignore_trivial=True, normalize=True, p_norm=2):
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/stump.py]: added to latest commit

File: stumpy/aamp.py [hunk 5]
```diff
@@ -224,11 +238,13 @@ def _aamp(T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_t
             if P[0, i, 2] > P[thread_idx, i, 2]:
                 P[0, i, 2] = P[thread_idx, i, 2]
                 I[0, i, 2] = I[thread_idx, i, 2]
-
-    return np.sqrt(P[0, :, :]), I[0, :, :]
+    if p_norm == 2:
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 6]
```diff
@@ -95,12 +96,15 @@ def _compute_diagonal(
...
+                    np.linalg.norm(T_B[i + k : i + k + m] - T_A[i : i + m], ord=p_norm)
+                    ** p_norm
+                )
             else:
                 D_squared = np.abs(
                     D_squared
-                    - (T_B[i + k - 1] - T_A[i - 1]) ** 2
-                    + (T_B[i + k + m - 1] - T_A[i + m - 1]) ** 2
+                    - np.absolute(T_B[i + k - 1] - T_A[i - 1]) ** p_norm
+                    + np.absolute(T_B[i + k + m - 1] - T_A[i + m - 1]) ** p_norm
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 7]
```diff
@@ -31,6 +31,7 @@ def _compute_diagonal(
     P,
     I,
     ignore_trivial,
+    p_norm=2,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 8]
```diff
@@ -95,12 +96,15 @@ def _compute_diagonal(
 
         for i in iter_range:
             if i == 0 or (k < 0 and i == -k):
-                D_squared = np.linalg.norm(T_B[i + k : i + k + m] - T_A[i : i + m]) ** 2
+                D_squared = (
+                    np.linalg.norm(T_B[i + k : i + k + m] - T_A[i : i + m], ord=p_norm)
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/stump.py [hunk 9]
```diff
@@ -42,6 +42,7 @@ def _compute_diagonal(
     ρ,
     I,
     ignore_trivial,
+    p_norm=2,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/stump.py]: added to latest commit

File: stumpy/aamp.py [hunk 10]
```diff
@@ -224,11 +238,13 @@ def _aamp(T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_t
             if P[0, i, 2] > P[thread_idx, i, 2]:
                 P[0, i, 2] = P[thread_idx, i, 2]
                 I[0, i, 2] = I[thread_idx, i, 2]
-
-    return np.sqrt(P[0, :, :]), I[0, :, :]
+    if p_norm == 2:
+        return np.sqrt(P[0, :, :]), I[0, :, :]
+    else:
+        return np.power(P[0, :, :], 1 / (p_norm)), I[0, :, :]
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 11]
```diff
@@ -136,7 +140,16 @@ def _compute_diagonal(
...
-def _aamp(T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_trivial):
+def _aamp(
+    T_A,
+    T_B,
+    m,
+    T_A_subseq_isfinite,
+    T_B_subseq_isfinite,
+    diags,
+    ignore_trivial,
+    p_norm=2,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 12]
```diff
@@ -208,6 +221,7 @@ def _aamp(T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_t
             P,
             I,
             ignore_trivial,
+            p_norm,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 13]
```diff
@@ -224,11 +238,13 @@ def _aamp(T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_t
...
-
-    return np.sqrt(P[0, :, :]), I[0, :, :]
+    if p_norm == 2:
+        return np.sqrt(P[0, :, :]), I[0, :, :]
+    else:
+        return np.power(P[0, :, :], 1 / (p_norm)), I[0, :, :]
 
 
-def aamp(T_A, m, T_B=None, ignore_trivial=True):
+def aamp(T_A, m, T_B=None, ignore_trivial=True, p_norm=2):
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/aamp.py [hunk 14]
```diff
@@ -303,7 +319,14 @@ def aamp(T_A, m, T_B=None, ignore_trivial=True):
...
     P, I = _aamp(
-        T_A, T_B, m, T_A_subseq_isfinite, T_B_subseq_isfinite, diags, ignore_trivial
+        T_A,
+        T_B,
+        m,
+        T_A_subseq_isfinite,
+        T_B_subseq_isfinite,
+        diags,
+        ignore_trivial,
+        p_norm,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/aamp.py]: added to latest commit

File: stumpy/stump.py [hunk 15]
```diff
@@ -235,6 +236,7 @@ def _stump(
     T_B_subseq_isconstant,
     diags,
     ignore_trivial,
+    p_norm=2,
```
[the remark this answers is not in the document]
Comment (vaish-muk) [reply stumpy/stump.py]: added to latest commit

**Comment by vaish-muk** [comment]
> @vaish-muk Overall this is a good start. In order for this PR to be complete, p-norm also needs to be added to:
> 
> `ammped.py` `gpu_aamp.py` `maamp.py`
> 
> along with their corresponding unit tests in order to ensure 100% code coverage. If it is not possible to add it to all of these places then we should avoid adding this feature altogether.

- p-norm added to aamped.py as part of latest commit

- gpu_aamp.py seems to be based on STOMP algorithm https://www.cs.ucr.edu/~eamonn/STOMP_GPU_final_submission_camera_ready.pdf This seems to be specific to euclidian distance since I see we are leveraging the dot products and based on D = T_B_subseq_squared[i] + T_A_subseq_squared[j] - 2.0 * QT_out[j]
- maamp also seems to have similar logic and refers to mSTOMP 

Please correct me if I am wrong about my understanding on gpu_aamp and maamp and help me with suggestions on how we can extend gpu_aamp and maamp as well to p-norm similar to aamp.


If you find it apt, we can merge the changes for extension to p-norm for aamp and aamped as part of this PR and work on gpu_aamp and maamp after further discussion as part of a fresh PR to avoid confusion.

File: stumpy/aamp.py [hunk 16]
```diff
@@ -95,12 +99,14 @@ def _compute_diagonal(
 
         for i in iter_range:
             if i == 0 or (k < 0 and i == -k):
-                D_squared = np.linalg.norm(T_B[i + k : i + k + m] - T_A[i : i + m]) ** 2
+                D_squared = (
```
Comment (BBischof) [inline stumpy/aamp.py:102]: the variable name is now questionable 😅

Comment (seanlaw) [reply stumpy/aamp.py:102]: Indeed 😄

**Comment by seanlaw** [comment]
@vaish-muk It looks like Github Actions just updated the `black` version (for checking code formatting) to `22.1.0`. This is causing your test to fail. Can you please:

1. Pull the latest changes to STUMPY and merge it with your branch
2. Update your local version of `black` to `22.1.0`

Thank you!

**Comment by seanlaw** [comment]
@vaish-muk Thank you for the new commits. Please allow me some time to review and formulate a proper response to your questions.

File: stumpy/aamp.py (hunk 16 above)
Comment (seanlaw) [reply stumpy/aamp.py:102]: @BBischof Any thoughts on what might be a more appropriate name?? `D_pow`??

Comment (BBischof) [reply stumpy/aamp.py:102]: I think keeping it simple and using `diagonal_p_norm`. Below it's that sliding window so there may be some additional language that helps communicate its use.

**Comment by seanlaw** [comment]
> In order for this PR to be complete, p-norm also needs to be added to:
> `ammped.py`
> `gpu_aamp.py`
> `maamp.py`

@vaish-muk I also forgot to add `aampi.py` to the list too 😢

**Comment by seanlaw** [comment]
> Please correct me if I am wrong about my understanding on gpu_aamp and maamp and help me with suggestions on how we can extend gpu_aamp and maamp as well to p-norm similar to aamp.

@vaish-muk I think the first thing that we should do is modernize all of the `aamp`-dependent modules so that they no longer depend on the sliding dot product `QT` for computing the distance. I've started a [new issue](https://github.com/TDAmeritrade/stumpy/issues/533) for this. Would you be willing to tackle this (with my assistance)?

> If you find it apt, we can merge the changes for extension to p-norm for aamp and aamped as part of this PR and work on gpu_aamp and maamp after further discussion as part of a fresh PR to avoid confusion.

I think it is best to modernize all of the `aamp`-related functions first in #533. Then, once that effort is complete, then we can come back and focus on this PR. How does that sound? I want to avoid having only partial support for p-norm and, instead, IMHO, it should be all or nothing.

**Comment by vaish-muk** [comment]
> > Please correct me if I am wrong about my understanding on gpu_aamp and maamp and help me with suggestions on how we can extend gpu_aamp and maamp as well to p-norm similar to aamp.
> 
> @vaish-muk I think the first thing that we should do is modernize all of the `aamp`-dependent modules so that they no longer depend on the sliding dot product `QT` for computing the distance. I've started a [new issue](https://github.com/TDAmeritrade/stumpy/issues/533) for this. Would you be willing to tackle this (with my assistance)?
> 
> > If you find it apt, we can merge the changes for extension to p-norm for aamp and aamped as part of this PR and work on gpu_aamp and maamp after further discussion as part of a fresh PR to avoid confusion.
> 
> I think it is best to modernize all of the `aamp`-related functions first in #533. Then, once that effort is complete, then we can come back and focus on this PR. How does that sound? I want to avoid having only partial support for p-norm and, instead, IMHO, it should be all or nothing.

Sure, we can work together on how to modernize those modules that work on sliding dot product QT as part of the new issue created https://github.com/TDAmeritrade/stumpy/issues/533 and then append those changes to this PR in addition to the ones implemented in this PR i.e  aamp and aamped

**Comment by vaish-muk** [comment]
committed black version related changes, should pass checks now

**Comment by seanlaw** [comment]
@vaish-muk I've been migrating all `aamp` non-normalized functions over to eliminate its dependency on `QT` and, in the process, also added `p-norm`. It's a big task and needing to touch nearly every file in the STUMPY repo. I've incorporated the wonderful changes that you've made here over into my branch. Thank you! Would you mind if I closed this PR in lieu of this development?

**Comment by vaish-muk** [comment]
Sure. But it would be even better if you could somehow include my PR/commit in the contributions to stumpy. So that I am officially listed as a contributor. 

Would love to continue to collaborate and contribute to stumpy in future as well

**Comment by seanlaw** [comment]
> Sure. But it would be even better if you could somehow include my PR/commit in the contributions to stumpy. So that I am officially listed as a contributor.
> 
> Would love to continue to collaborate and contribute to stumpy in future as well

Sounds good. Let me review, merge, and then I'll push my commits after this is merged!

**Comment by vaish-muk** [comment]
Thanks @seanlaw that'll be great!

**Comment by seanlaw** [comment]
Thank you for your contribution @vaish-muk! 🎉 We look forward to collaborating with you again 👍

**Comment by vaish-muk** [comment]
> Thank you for your contribution @vaish-muk! 🎉 We look forward to collaborating with you again 👍

Same here, looking forward to future collaborations 😊

**Pull request merged**